sparse/sfold.go

```diff
@@ -63,36 +53,16 @@ func FoldFile(childFileName, parentFileName string, ops FoldFileOperations) erro
 	return coalesce(parentFileIo, childFileIo, childFInfo.Size(), ops)
 }
 
-func coalesce(parentFileIo, childFileIo FileIoProcessor, fileSize int64, ops FoldFileOperations) (err error) {
-	var progress uint32
+func coalesce(parentFileIo, childFileIo FileIoProcessor, fileSize int64, ops FileHandlingOperations) (err error) {
+	progress := new(uint32)
 	progressMutex := &sync.Mutex{}
 
-	// progress updates can only be monotonically increasing, once that's replaced we can use an atomic cmp&swp around progress
```

Comment:
Actually, I am not sure if we could replace CAS with the locking. Because CAS could handle variable `progress` update only but we need to make sure the variable update **and the interface call `ops.UpdateFoldFileProgress()`** atomic. 

Here is the replacement I tried but I don't think it would work:
```
func updateProgressCAS(currentProgressAddr *uint32, newProgress uint32, done bool, err error, progressMutex *sync.Mutex, ops FileHandlingOperations) {
	forcedUpdate := done || err != nil
	if forcedUpdate {
		ops.UpdateFileHandlingProgress(int(newProgress), done, err)
		atomic.StoreUint32(currentProgressAddr, newProgress)
		return
	}

	for currentProgress := atomic.LoadUint32(currentProgressAddr); currentProgress < newProgress; currentProgress = atomic.LoadUint32(currentProgressAddr) {
		if atomic.CompareAndSwapUint32(currentProgressAddr, currentProgress, newProgress) {
			ops.UpdateFileHandlingProgress(int(newProgress), done, err)
			return
		}
	}
}
```

Reply:
atomic-related functions only apply to the target pointer so outside of this scope, it can't control. As you said, the example here should not work.

Reply:
Then can we remove this comment?
cc @joshimoo

Reply:
Unlike file syncing, there is no receiver or sender during coalescing/pruning.
Or do you mean that using a separate goroutine with a channel as the receiver? The channel is actually using locking to protect the signal transmission.